### PR TITLE
Avoid un-necessary unwrapping of Tensor in SavedVariable

### LINKED ISSUE #58500: Avoid un-necessary unwrapping of Tensor in SavedVariable
SavedVariable is current doing unwrapping as:
- For leaf, save a weak pointer to its grad-accumulator
- If the Tensor is an output of the function, save a weak pointer to its grad_fn
- If the Tensor is none of the above, save a strong pointer to its grad_fn

On top of that, it also unpacks the version counter, forward gradients, requires grad field and a `.data` of the Tensor itself.
This allows the SavedVariable to reconstruct a similar Tensor when it is requested during the backward.

The first and third categories here though don't actually need to go through the whole unwrapping and the given Tensor can be kept as-is (the saved version and data reset logic will need to remain of course).

This task would consist in updating this logic to keep the original Tensor as is in these cases and add corresponding tests.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

### PR DESCRIPTION
**Pull request by Varal7**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59838 Remove dead code in SavedVariable
* **#59837 Avoid un-necessary unwrapping of Tensor in SavedVariable**

Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069215](https://our.internmc.facebook.com/intern/diff/D29069215)

### CODE REVIEW

File: torch/csrc/autograd/saved_variable.cpp [hunk 1]
```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     version_counter_ = impl::version_counter(variable);
     saved_version_ = version_counter_.current_version();
 
+    // This check actually does not hold
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:44]: Why? When you consider inplace ops?

File: torch/csrc/autograd/saved_variable.cpp [hunk 2]
```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
...
+    // TORCH_CHECK(is_output && variable.is_leaf(), "Variable is both an output and a leaf");
+
+    // If the variable is not an output, we can safely save the
+    // original variable without running the risk of reference cycles.
+    // If the variable is not an output, its grad_fn has already been fully
+    // created and in particular will be a different Node than the one
+    // we are currently constructing (the one that owns this SavedVariable).
+    // Similarly, if the variable is a leaf.
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:52]: I think the leaf case is different actually. You don't have a cycle because the Tensor only has a weak ref to the grad_accumulator. So that cannot create a cycle.

File: torch/csrc/autograd/saved_variable.h [hunk 3]
```diff
@@ -70,5 +74,6 @@ class TORCH_API SavedVariable {
   bool requires_grad_ = false;
   bool has_grad_fn_ = false;
   bool is_inplace_view_ = false;
+  bool saved_original = false;
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.h:77]: Use `_` at the end for non-public attributes

File: torch/csrc/autograd/saved_variable.h [hunk 4]
```diff
@@ -46,6 +46,10 @@ class TORCH_API SavedVariable {
   }
 
  private:
+  // The variable to save.
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.h:49]: The doc here should mention the link between the content of this and the boolean saved_original below.

File: torch/csrc/autograd/saved_variable.cpp [hunk 5]
```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     version_counter_ = impl::version_counter(variable);
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:41]: Why do you still save the version counter here?

File: test/test_autograd.py [hunk 6]
```diff
@@ -5421,6 +5421,26 @@ def get_refs(with_backward):
...
+
+        # b is not an output
+        self.assertIs(c.grad_fn._saved_other, b)
+
+        c.sum().backward()
+
+        with self.assertRaisesRegex(RuntimeError, "after they have already been freed"):
+            c.grad_fn._saved_self
+
+        self.assertEqual(a, torch.ones(5))
```
Comment (albanD) [inline test/test_autograd.py:5442]: Why do you check the value of a here? Did you meant to check the value of a.grad? If so, maybe use another value than 1 for a so that you get non-1 gradients.

File: test/test_autograd.py [hunk 7]
```diff
@@ -7332,7 +7352,9 @@ def test_sparse_backward(self, device, dtype):
         class FixedGradientFunction(Function):
             @staticmethod
             def forward(ctx, x, grad_x):
-                ctx.save_for_backward(grad_x)
+                # this is needed because of https://github.com/pytorch/pytorch/issues/59828
```
Comment (albanD) [inline test/test_autograd.py:7355]: The fix landed now. So after rebasing on top of master, you should be able to remove this.

**Review by albanD — COMMENTED** [review]
Looks ok to me.
Is there any other tests that we want to do here?

**Comment by Varal7** [comment]
> Looks ok to me.
> Is there any other tests that we want to do here?

Thanks for the review! I will address your comments inline.
I'm happy with the current tests (after I fix the one you pointed out)

File: torch/csrc/autograd/saved_variable.cpp (hunk 5 above)
Comment (Varal7) [reply torch/csrc/autograd/saved_variable.cpp:41]: Yes I think you're right that I don't need it here.

File: test/test_autograd.py (hunk 6 above)
Comment (Varal7) [reply test/test_autograd.py:5442]: I meant to check that the value of `a` has not been cleared by `release_variables`.

Comment (albanD) [reply test/test_autograd.py:5442]: Ah! Maybe leave a comment there "a is still untouched"?
Also, unless we do something very bad with refcounting, there is no reason for a to have disapeared here.

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).

File: test/test_autograd.py [hunk 8]
```diff
@@ -5467,6 +5467,28 @@ def test_input_buffer_accum(self):
         # Given gradients should not be modified inplace
         self.assertEqual(grad_out1, grad_out1_original)
 
+    def test_no_unnecessary_unwrapping(self):
+        a = torch.randn(5, requires_grad=True)
+        a_orig = a.detach().clone()
+        b = a * a
+        c = a * b
```
Comment (albanD) [inline test/test_autograd.py:5474]: nit: if you want to test an output here to make sure it is properly unwrap, you can use exp() and make sure the `_saved_result` is not the same as the returned Tensor.

File: torch/csrc/autograd/saved_variable.cpp [hunk 9]
```diff
@@ -37,13 +37,30 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
       "you can make a clone to get a normal tensor and use it in autograd.")
 
     was_default_constructed_ = false;
-    is_inplace_view_ = is_inplace_view;
-    version_counter_ = impl::version_counter(variable);
-    saved_version_ = version_counter_.current_version();
+    is_inplace_on_view_ = is_inplace_on_view;
+    auto& version_counter = impl::version_counter(variable);
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:41]: Why not `const auto&` ?

File: torch/csrc/autograd/saved_variable.cpp [hunk 10]
```diff
@@ -37,13 +37,30 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
...
+    saved_version_ = version_counter.current_version();
+
+    // This check actually does not hold
+    // TORCH_CHECK(is_output && variable.is_leaf(), "Variable is both an output and a leaf");
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:45]: This commented code should be removed.

File: torch/csrc/autograd/saved_variable.h [hunk 11]
```diff
@@ -46,6 +46,15 @@ class TORCH_API SavedVariable {
   }
 
  private:
+  // This field contains either:
+  // 1. the variable to save
+  // 2. or its tensor_data.
+  // If storing the variable itself would create a circular reference,
+  // we fall into the second case and its metadata is also saved separately.
+  // In that case, the grad_fn must be passed in to the unpack function when
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.h:54]: nit: not true for inplace_on_view case.

**Review by albanD — APPROVED** [review]
LGTM
Minor updates to remove spurious code.

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).
